File: python/src/auditable_mcp/research_tool.py
"""Dummy first-party MCP tool (research assistant).

Demonstrates tool-internal domain operations across the (mutates, egress) axis.
Note that web searches are read-only (mutates=False) but leak query parameters (egress=True).
"""

from auditable_mcp.amcp import AmcpSession

SEARCH_ENDPOINT = 'https://api.search.example/v1/search'


class ResearchTool:
    """An in-memory note store plus a web search, all self-attested."""

    def __init__(self, session: AmcpSession) -> None:
        """Bind the audit session and start with no notes."""
        self._session = session
        self._notes: dict[str, dict] = {}

    def search(self, query: str) -> list[dict]:
        """Search the web (api.request).

        Mutates nothing, but the query egresses. The query is hashed into params_hash and
        never stored raw: it is precisely the sensitive part that leaked.
        """
        return self._session.audited(
            action_type='api.request',
            target_resource={'kind': 'endpoint', 'ref': SEARCH_ENDPOINT},
            params={'query': query},
            perform=lambda: [{'title': f'Result for {query}', 'url': 'https://example.com/a'}],
            mutates=False,
            egress=True,
        )

    def save_note(self, topic: str, content: str) -> dict:
        """Persist a note (db.write) -- a state change that stays inside the trust boundary."""

        def perform() -> dict:
            note = {'topic': topic, 'content': content}
            self._notes[topic] = note
            return note

        return self._session.audited(
            action_type='db.write',
            target_resource={'kind': 'table', 'ref': 'notes', 'scope_hint': f'topic={topic}'},
            params={'topic': topic, 'content': content},
            perform=perform,
            mutates=True,
            egress=False,
        )

    def list_notes(self) -> list[dict]:
        """List notes (db.read) -- the genuinely benign case: no state change, no egress."""
        return self._session.audited(
            action_type='db.read',
            target_resource={'kind': 'table', 'ref': 'notes'},
            params={},
            perform=lambda: list(self._notes.values()),
            mutates=False,
            egress=False,
        )

File: reference/typescript/src/tool/researchTool.ts
import type { AmcpSession } from './amcp.js';

// Dummy first-party MCP tool (research assistant). Its three operations span the
// (mutates, egress) axis. A web search is read-only (mutates=false) but leaks its query
// parameters (egress=true).

export interface Note {
  topic: string;
  content: string;
}

export interface SearchHit {
  title: string;
  url: string;
}

const SEARCH_ENDPOINT = 'https://api.search.example/v1/search';

export class ResearchTool {
  private readonly notes = new Map<string, Note>();

  constructor(private readonly session: AmcpSession) {}

  // api.request: mutates nothing, but the query egresses. The query is hashed into
  // params_hash, never stored raw.
  async search(query: string): Promise<SearchHit[]> {
    return this.session.audited(
      {
        action_type: 'api.request',
        target_resource: { kind: 'endpoint', ref: SEARCH_ENDPOINT },
        params: { query },
        effect: { mutates: false, egress: true },
      },
      async () => [{ title: `Result for ${query}`, url: 'https://example.com/a' }],
    );
  }

  // db.write: a state change inside the trust boundary.
  async saveNote(topic: string, content: string): Promise<Note> {
    return this.session.audited(
      {
        action_type: 'db.write',
        target_resource: { kind: 'table', ref: 'notes', scope_hint: `topic=${topic}` },
        params: { topic, content },
        effect: { mutates: true, egress: false },
      },
      async () => {
        const note: Note = { topic, content };
        this.notes.set(topic, note);
        return note;
      },
    );
  }

  // db.read: no state change, no egress.
  async listNotes(): Promise<Note[]> {
    return this.session.audited(
      {
        action_type: 'db.read',
        target_resource: { kind: 'table', ref: 'notes' },
        params: {},
        effect: { mutates: false, egress: false },
      },
      async () => [...this.notes.values()],
    );
  }
}
